fsw/sensor-fw/src/adc.rs

```diff
@@ -0,0 +1,309 @@
+use core::ops::Deref;
+
+use defmt::Format;
+use hal::pac;
+
+const VREFINT_CAL_ADDR: usize = 0x1FF1E860;
+const TS_CAL1_ADDR: usize = 0x1FF1E820;
+const TS_CAL2_ADDR: usize = 0x1FF1E840;
+const VREFINT_CHANNEL: u8 = 19;
+const VSENSE_CHANNEL: u8 = 18;
+const VBAT_CHANNEL: u8 = 17;
+const TS_CAL1_TEMP: f32 = 30.0;
+const TS_CAL2_TEMP: f32 = 110.0;
+
+/// Safe interface for reading factory calibration values from memory
+///
+/// # Safety
+/// This function reads from well-defined factory calibration memory addresses
+/// that are guaranteed to be valid on STM32H7 microcontrollers. These addresses
+/// contain factory-programmed calibration data and are safe to read.
+fn read_factory_u16(addr: usize) -> u16 {
```

Comment:
I think this function shouldn't be considered safe because it could be used in an unsafe way,. In the code in its current state it is safe, but you usually mark a function unsafe if it can be potentially used in an unsafe way. I would also make it take the address as a const ptr, then in the const defs you can cast the usizes to the ptrs.